File: programmers/kakao/level3/셔틀버스.py
def solution(n, t, m, timetable):
    timetable = sorted([int(time[:2])*60+int(time[3:]) for time in timetable])

    busTime = [540+t*i for i in range(n)]

    i = 0
    for bt in busTime:
        cnt = 0
        while cnt < m and i < len(timetable) and timetable[i] <= bt:
            i += 1
            cnt += 1
        if cnt < m:
            answer = bt
        else:
            answer = timetable[i-1]-1

    return str(answer//60).zfill(2)+":"+str(answer%60).zfill(2)


n, t, m = 1, 1, 5
timetable = ["08:00", "08:01", "08:02", "08:03"]
print(solution(n, t, m, timetable)) # "09:00"

n, t, m = 2, 10, 2
timetable = ["09:10", "09:09", "08:00"]
print(solution(n, t, m, timetable)) # "09:09"

n, t, m = 2, 1, 2
timetable = ["09:00", "09:00", "09:00", "09:00"]
print(solution(n, t, m, timetable)) # "08:59"

n, t, m = 1, 1, 5
timetable = ["00:01", "00:01", "00:01", "00:01", "00:01"]
print(solution(n, t, m, timetable)) # "00:00"

n, t, m = 1, 1, 1
timetable = ["23:59"]
print(solution(n, t, m, timetable)) # "09:00"

n, t, m = 10, 60, 45
timetable = ["23:59","23:59", "23:59", "23:59", "23:59", "23:59", "23:59", "23:59", "23:59", "23:59", "23:59", "23:59", "23:59", "23:59", "23:59", "23:59"]
print(solution(n, t, m, timetable)) # "18:00"

File: Java/programmers/kakao/level3/셔틀버스.java
package CodeTest.Java.programmers.kakao.level3;

import java.util.*;

public class 셔틀버스 {
    public String solution(int n, int t, int m, String[] timetable) {
        List<Integer> table = new ArrayList<>();
        for (String time : timetable) {
            table.add(Integer.parseInt(time.substring(0, 2))*60 + Integer.parseInt(time.substring(3)));
        }
        Collections.sort(table);
        
        int[] busTime = new int[n];
        for (int i=0; i<n; i++) {
            busTime[i] = 540+t*i;
        }
        
        int i = 0;
        int answer = 0;
        for (int bt : busTime) {
            int cnt = 0;
            while (cnt < m && i < table.size() && table.get(i) <= bt) {
                i++;
                cnt++;
            }
            if (cnt < m) answer = bt;
            else answer = table.get(i-1)-1;
        }
        
        return String.format("%02d:%02d", answer/60, answer%60);
    }
}
